platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java

```diff
@@ -20,33 +20,37 @@
 import java.util.List;
 import java.util.concurrent.locks.StampedLock;
 import org.apache.felix.cm.PersistenceManager;
-import org.codice.felix.cm.internal.ConfigurationContextFactory;
 import org.codice.felix.cm.internal.ConfigurationPersistencePlugin;
 import org.osgi.framework.BundleContext;
 import org.osgi.framework.FrameworkUtil;
-import org.osgi.framework.ServiceReference;
 import org.osgi.util.tracker.ServiceTracker;
-import org.osgi.util.tracker.ServiceTrackerCustomizer;
 
 /**
  * The {@link DelegatingPersistenceManager} is responsible for iterating over relevant configuration
  * plugin points prior to moving the configs to the inner {@link PersistenceManager}.
  *
+ * <p>Read locks are maintained during configuration persistence and removal so that initializing
+ * plugins do not miss configuration data.
+ *
  * <p><b>See FELIX-4005 & FELIX-4556. This class cannot utilize Java 8 language constructs due to
  * maven bundle plugin 2.3.7</b>
+ *
+ * @see ConfigPluginCustomizer
  */
 public class DelegatingPersistenceManager extends WrappedPersistenceManager {
   private final ServiceTracker<ConfigurationPersistencePlugin, ConfigurationPersistencePlugin>
       configPersistenceTracker;
 
-  public DelegatingPersistenceManager(
-      PersistenceManager persistenceManager, StampedLock configDeadZoneLock) {
-    this(
-        persistenceManager,
+  private final StampedLock serviceStartingLock = new StampedLock();
+
+  public DelegatingPersistenceManager(PersistenceManager persistenceManager) {
+    super(persistenceManager);
+    this.configPersistenceTracker =
         new ServiceTracker<>(
             getBundleContext(),
             ConfigurationPersistencePlugin.class,
-            new ConfigInterceptingCustomizer(configDeadZoneLock)));
+            new ConfigPluginCustomizer(getInnerPersistenceManager(), serviceStartingLock));
```

Comment:
❗️ Could we avoid passing the lock outside of this class and keep everything internal instead? 

If not then I would really suggest not exposing what type of lock it is. The plugin customizer class doesn't need to know why it has to grab a write lock instead of a read lock. This is very specific to the implementation if the DelegatingPersistenceManager class.

My preference would be to keep that class responsible for detecting new plugins and call this delegating persistence manager when it does such that this persistence manager is responsible for all config objects related operations and in this case, it would be to grab the lock, get all configs from the inner persistence manager and call the init for that new service added. This keeps all interactions with the services listening for events and all cache interactions (read, update, delete) in one place; here.